src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation. For a given function specified by the input
+ * global var, it generates a new function which computes the adjoints of the given function's
+ * inputs in respect of the function's output. Use require_grads to specify inputs which need to be
+ * differentiated instead of returning adjoints of all inputs.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(std::move(mod)), require_grads_(std::move(require_grads)) {}
+
+  Function FuncTransform(Function func) {
+    ICHECK(func->body->IsInstance<SeqExprNode>());
+
+    Array<Var> new_params;
+    for (Var param : func->params) {
+      Var new_param = Var(param->vid, param->shape(), param->checked_type(), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+    Expr new_body = VisitWithNewScope(func->body);
+
+    const SeqExprNode* seq_expr = new_body.as<SeqExprNode>();
+    // only support a single dataflow block
+    ICHECK(seq_expr->blocks.size() == 1);
+    ICHECK(seq_expr->blocks[0]->IsInstance<DataflowBlockNode>());
+    const DataflowBlockNode* block = seq_expr->blocks[0].as<DataflowBlockNode>();
+
+    builder_->BeginDataflowBlock();
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+
+    // create adjoint var for inputs
+    for (size_t i = 0; i < new_params.size(); ++i) {
+      if (require_grads_.empty() || std::find(require_grads_.begin(), require_grads_.end(),
+                                              func->params[i]) != require_grads_.end()) {
```

Comment:
Hmmm, `std::find(...) != ..end()` means we find the param in `require_grads_`, which is conflict with `require_grads_.empy()`. Are you sure this is intended? And if so, when `require_grads_.empty()` is parallel with `std::find(...) != require_grads_.end()`?

Reply:
The logic here is that if `require_grads_.empty()`, which means default, we do AD to all inputs

Reply:
If not the default case, meaning that the user specified some variables in `require_grads_`, we only do AD to variables in it.

Reply:
I see. It is very easy for people to think “if `require_grads_` is empty then we compute gradients from no one”. So one way is to make the `require_grads` an `Optional<Array<Var>>` in the interface: being NullOpt means the default case, where we will compute adjoints for every param; otherwise, we compute the adjoints only for the Vars in this array.

And in the following function
https://github.com/mlc-ai/relax/blob/75081dbda4f0d8ac4e4c3a826bf60294e1b45989/src/relax/transform/gradient.cc#L373-L378
once you find `require_grads` is NullOpt, you can set it to the list of Function params. This can make the logic clearer and less error-prone.

Reply:
Sounds reasonable